File: maps/src/main/java/it/localhostsoftware/maps/cameraUpdateFactory/CameraUpdateFactory.java
package it.localhostsoftware.maps.cameraUpdateFactory;

import android.content.Context;
import android.graphics.Point;

import com.google.android.gms.common.ConnectionResult;
import com.google.android.gms.common.GoogleApiAvailability;
import com.google.android.gms.maps.model.CameraPosition;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.LatLngBounds;

import it.localhostsoftware.maps.CameraUpdate;

public interface CameraUpdateFactory {
    static CameraUpdateFactory getInstance(Context context) {
        if (GoogleApiAvailability.getInstance().isGooglePlayServicesAvailable(context) == ConnectionResult.SUCCESS)
            return new GoogleCameraUpdateFactory();
        else return null;
    }

    CameraUpdate zoomIn();

    CameraUpdate zoomOut();

    CameraUpdate scrollBy(float var0, float var1);

    CameraUpdate zoomTo(float var0);

    CameraUpdate zoomBy(float var0);

    CameraUpdate zoomBy(float var0, Point var1);

    CameraUpdate newCameraPosition(CameraPosition var0);

    CameraUpdate newLatLng(LatLng var0);

    CameraUpdate newLatLngZoom(LatLng var0, float var1);

    CameraUpdate newLatLngBounds(LatLngBounds var0, int var1);

    CameraUpdate newLatLngBounds(LatLngBounds var0, int var1, int var2, int var3);
}

File: maps/src/main/java/it/localhostsoftware/maps/CameraUpdateFactory.kt
package it.localhostsoftware.maps

import android.content.Context
import android.graphics.Point
import it.localhostsoftware.maps.google.GoogleCameraUpdateFactory
import it.localhostsoftware.maps.huawei.HuaweiCameraUpdateFactory
import it.localhostsoftware.maps.model.CameraPosition
import it.localhostsoftware.maps.model.LatLng
import it.localhostsoftware.maps.model.LatLngBounds

interface CameraUpdateFactory<CU : CameraUpdate<*>, CP : CameraPosition<*, *>, LL : LatLng<*>, LLB : LatLngBounds<*, *>> {
    companion object {
        fun getInstance(c: Context) =
            when (c.getMobileServices()) {
                MobileServices.GOOGLE -> GoogleCameraUpdateFactory()
                MobileServices.HUAWEI -> HuaweiCameraUpdateFactory()
                else -> throw IllegalStateException()
            }
    }

    fun zoomIn(): CU
    fun zoomOut(): CU
    fun scrollBy(var0: Float, var1: Float): CU
    fun zoomTo(var0: Float): CU
    fun zoomBy(var0: Float): CU
    fun zoomBy(var0: Float, var1: Point): CU
    fun newCameraPosition(var0: CP): CU
    fun newLatLng(var0: LL): CU
    fun newLatLngZoom(var0: LL, var1: Float): CU
    fun newLatLngBounds(var0: LLB, var1: Int): CU
    fun newLatLngBounds(var0: LLB, var1: Int, var2: Int, var3: Int): CU
}